pkg/controller/metrics/clientwrapper.go

```diff
@@ -0,0 +1,104 @@
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	metricKubeClientRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "hive_kube_client_requests_total",
+		Help: "Counter incremented for each kube client request.",
+	},
+		[]string{"controller", "method", "resource"},
+	)
+)
+
+func init() {
+	metrics.Registry.MustRegister(metricKubeClientRequests)
+}
+
+// NewClientWithMetricsOrDie creates a new controller-runtime client with a wrapper which increments
+// metrics for requests by controller name, HTTP method, and URL path. The client will re-use the
+// managers cache. This should be used in all Hive controllers.
+func NewClientWithMetricsOrDie(mgr manager.Manager, ctrlrName string) client.Client {
+	// Copy the rest config as we want our round trippers to be controller specific.
+	cfg := rest.CopyConfig(mgr.GetConfig())
+	cfg.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+		return &ControllerMetricsTripper{
+			RoundTripper: rt,
+			controller:   ctrlrName,
+		}
+	}
+
+	options := client.Options{
+		Scheme: mgr.GetScheme(),
+		Mapper: mgr.GetRESTMapper(),
+	}
+	c, err := client.New(cfg, options)
+	if err != nil {
+		log.WithError(err).Fatal("unable to initialize metrics wrapped client")
```

Comment:
We are not returning from here in case of error. Is this as designed?

Reply:
Yeah log.Fatal includes a sys.Exit, so we're exiting the program if something goes wrong here as it's controller initialization. This is why I went with the "OrDie" in the function name.

Reply:
Got it. Thanks.